src/key_io.cpp

```diff
@@ -309,3 +311,33 @@ bool IsValidDestinationString(const std::string& str)
 {
     return IsValidDestinationString(str, Params());
 }
+
+std::pair<CPubKey, CPubKey> DecodeSilentData(const std::vector<unsigned char>& data)
+{
+
+    std::vector<unsigned char> scan_pubkey_data(data.begin(), data.begin() + 33);
+    CPubKey scan_pubkey{scan_pubkey_data};
+
+    std::vector<unsigned char> spend_pubkey_data(data.begin() + 33, data.end());
+    CPubKey spend_pubkey{spend_pubkey_data};
+
+    return {scan_pubkey, spend_pubkey};
+}
+
+std::vector<unsigned char> DecodeSilentAddress(const std::string& str)
+{
+    const auto& params{Params()};
+    const auto& silent_payment_hrp = params.SilentPaymentHRP();
+    const auto dec = bech32::Decode(str, /*silent=*/true);
+    if (dec.encoding != bech32::Encoding::BECH32M || dec.hrp != silent_payment_hrp) {
+        return {};
+    }
+    auto version = dec.data.front();  // retrieve the version
+    std::vector<unsigned char> silent_payment_data;
+    silent_payment_data.reserve(((dec.data.size() - 1) * 5) / 8);
+    if (!ConvertBits<5, 8, false>([&](unsigned char c) { silent_payment_data.push_back(c); }, dec.data.begin() + 1, dec.data.end())) {
+        return {};
+    }
+    if ((version == 0 && silent_payment_data.size() != SILENT_PAYMENT_V0_DATA_SIZE) || silent_payment_data.size() < SILENT_PAYMENT_V0_DATA_SIZE) return {};
```

Comment:
37b0ca1c3d753cc61e0debf8ea59160bdf4127ec: could use `version == 0 && silent_payment_data.size() < SILENT_PAYMENT_V0_DATA_SIZE` (see comment above)

Reply:
I think this should fail if an address says it is v0 but has a data part larger than what is defined for v0. If an address has a version > v0 and data part > V0_SIZE, then only the _first_ 66 bytes of the data part are read, which is what we want for the forward compatibility bit.